### DISCUSSION TITLE
Where can I find the list of return codes and how to fix them?

### DISCUSSION THREAD

**Discussion by tukusejssirs** [Ladder99/fanuc-driver#93] [Q&A]
@MRIIOT, you have once said that to fix `{"method": "cnc_rdopmsg", "rc": 9}` error, one needs to set `11931#1` to `0`. It works, but I could not find what does this parameter do, why it is needed. I could not find a reference to this on inventcom.net.

Now, on some machine, I receive `{"method": "cnc_toolnum", "rc": 6}` error. Is there a fix for that?

It would be nice if the return codes (and possible ways to resolve them) were documented on docs.ladder99.com. I know the codes are defined in [`fanuc/fwlib.cs` (L53-L99)](https://github.com/Ladder99/fanuc-driver/blob/main/fanuc/fwlib.cs#L53-L99), but I am not sure if `EW_NOOPT = 6, /* no option error */` is anyhow related to `{"method": "cnc_toolnum", "rc": 6}` error I get.

**Comment by MRIIOT** [comment]
> you have once said that to fix {"method": "cnc_rdopmsg", "rc": 9} error, one needs to set 11931#1 to 0. It works, but I could not find what does this parameter do, why it is needed. I could not find a reference to this on inventcom.net.

You will find this in the Fanuc controller manual.  Not in Focas docs.  Why this parameter has an effect on operator messages is black Fanuc magic to me.

<img width="569" alt="image" src="https://github.com/Ladder99/fanuc-driver/assets/48079336/1911900e-2979-4f68-83a9-0c1716576ab4">
  
  

> {"method": "cnc_toolnum", "rc": 6}

I do not think there is a fix for this.  'no option' to me means that the controller does not support tool life management functions.  If `cnc_toolnum` fails, the fallback is this: https://github.com/Ladder99/fanuc-driver/blob/f90569022c24d1409a1352a8fd4acb9fe574fc9d/fanuc/collectors/ToolData.cs#L24-L25
